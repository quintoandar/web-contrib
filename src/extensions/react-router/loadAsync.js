export const errorLoading = (errorHandler) => (err) => {
  console.error('Dynamic page loading failed', err); // eslint-disable-line no-console
  if (
    process.env.NODE_ENV === 'production' &&
    typeof errorHandler === 'function'
  ) {
    errorHandler(err);
  }
};

export const loadModule = (cb) => (componentModule) => {
  cb(null, componentModule.default);
};

export default (errorHandler) => (importedComponent, cb) => {
  const renderRoute = loadModule(cb);
  return Promise.all([importedComponent])
    .then(([component]) => {
      renderRoute(component);
    })
    .catch(errorLoading(errorHandler));
};
