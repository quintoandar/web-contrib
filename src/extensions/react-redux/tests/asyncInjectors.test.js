import { fromJS } from 'immutable';
import { memoryHistory } from 'react-router';
import configureStore from 'extensions/redux/store';

import {
  injectAsyncReducer,
  injectAsyncFirestoreListener,
  getAsyncInjectors,
} from '../asyncInjectors';

// Fixtures

const initialState = fromJS({ reduced: 'soon' });

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'TEST':
      return state.set('reduced', action.payload);
    default:
      return state;
  }
};

const firestoreListener = (dispatch) => {
  dispatch({
    type: 'TEST_ACTION',
    data: { foo: 'jkbzrk' },
  });
};

describe('asyncInjectors', () => {
  let store;

  describe('getAsyncInjectors', () => {
    beforeAll(() => {
      store = configureStore({}, memoryHistory);
    });

    it('given a store, should return all async injectors', () => {
      const { injectReducer } = getAsyncInjectors(store);

      injectReducer('test', reducer);
      store.dispatch({ type: 'TEST', payload: 'yup' });

      const actual = store.getState().get('test');
      const expected = initialState.merge({ reduced: 'yup' });

      expect(actual.toJS()).toEqual(expected.toJS());
    });

    it('should throw if passed invalid store shape', () => {
      let result = false;

      delete store.dispatch;

      try {
        getAsyncInjectors(store);
      } catch (err) {
        result = err.name === 'Invariant Violation';
      }

      expect(result).toEqual(true);
    });
  });

  describe('helpers', () => {
    beforeAll(() => {
      store = configureStore({}, memoryHistory);
    });

    describe('injectAsyncReducer', () => {
      it('given a store, it should provide a function to inject a reducer', () => {
        const injectReducer = injectAsyncReducer(store);

        injectReducer('test', reducer);

        const actual = store.getState().get('test');
        const expected = initialState;

        expect(actual.toJS()).toEqual(expected.toJS());
      });

      it('should not assign reducer if already existing', () => {
        const injectReducer = injectAsyncReducer(store);

        injectReducer('test', reducer);
        injectReducer('test', () => {});

        expect(store.asyncReducers.test.toString()).toEqual(reducer.toString());
      });

      it('should throw if passed invalid name', () => {
        let result = false;

        const injectReducer = injectAsyncReducer(store);

        try {
          injectReducer('', reducer);
        } catch (err) {
          result = err.name === 'Invariant Violation';
        }

        try {
          injectReducer(999, reducer);
        } catch (err) {
          result = err.name === 'Invariant Violation';
        }

        expect(result).toEqual(true);
      });

      it('should throw if passed invalid reducer', () => {
        let result = false;

        const injectReducer = injectAsyncReducer(store);

        try {
          injectReducer('bad', 'nope');
        } catch (err) {
          result = err.name === 'Invariant Violation';
        }

        try {
          injectReducer('coolio', 12345);
        } catch (err) {
          result = err.name === 'Invariant Violation';
        }

        expect(result).toEqual(true);
      });
    });

    describe('injectAsyncFirestoreListener', () => {
      let listenerStore;
      let injectFirestoreListener;

      beforeEach(() => {
        jest.resetAllMocks();
        listenerStore = {
          dispatch: jest.fn(),
          asyncFirestoreListeners: [],
        };
        injectFirestoreListener = injectAsyncFirestoreListener(listenerStore);
      });

      it('given a store, it should provide a function to inject a firestore listener', () => {
        injectFirestoreListener('test', firestoreListener);

        expect(listenerStore.asyncFirestoreListeners.test).toBe(firestoreListener);
        expect(listenerStore.dispatch).toHaveBeenCalled();
      });

      it('should not assign listener if already existing', () => {
        injectFirestoreListener('test', firestoreListener);
        injectFirestoreListener('test', () => {});

        expect(listenerStore.asyncFirestoreListeners.test.toString())
          .toEqual(firestoreListener.toString());
      });

      it('should throw if passed invalid name - empty', () => {
        expect(() => injectFirestoreListener('', firestoreListener)).toThrowErrorMatchingSnapshot();
      });

      it('should throw if passed invalid name - number', () => {
        expect(() => injectFirestoreListener(999, firestoreListener)).toThrowErrorMatchingSnapshot();
      });

      it('should throw if passed invalid listener - number', () => {
        expect(() => injectFirestoreListener('coolio', 12345)).toThrowErrorMatchingSnapshot();
      });

      it('should throw if passed invalid listener - string', () => {
        expect(() => injectFirestoreListener('bad', 'nope')).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
