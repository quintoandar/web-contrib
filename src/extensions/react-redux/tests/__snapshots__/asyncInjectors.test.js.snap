// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncInjectors helpers injectAsyncFirestoreListener should throw if passed invalid listener - number 1`] = `"(app/utils...) injectFirestoreListener: Expected \`asyncListener\` to be a firestore listener function"`;

exports[`asyncInjectors helpers injectAsyncFirestoreListener should throw if passed invalid listener - string 1`] = `"(app/utils...) injectFirestoreListener: Expected \`asyncListener\` to be a firestore listener function"`;

exports[`asyncInjectors helpers injectAsyncFirestoreListener should throw if passed invalid name - empty 1`] = `"(app/utils...) injectFirestoreListener: Expected \`asyncListener\` to be a firestore listener function"`;

exports[`asyncInjectors helpers injectAsyncFirestoreListener should throw if passed invalid name - number 1`] = `"(app/utils...) injectFirestoreListener: Expected \`asyncListener\` to be a firestore listener function"`;
